File: PART 2/Chapter 02/Problem 06/3.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

var N = 0
var M = 0
var gr = Array(0) { Array(0) { false } }
var chk = Array(0) { false }
var ans = 0

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    BufferedWriter(OutputStreamWriter(System.out)).use { bw ->
        val nm = readLine().split(" ").map { it.toInt() }
        N = nm[0]
        M = nm[1]
        gr = Array(N) { Array(N) { false } }
        chk = Array(N) { false }
        repeat(M) {
            val (u, v) = readLine().split(" ").map { it.toInt() - 1 }
            gr[u][v] = true
            gr[v][u] = true
        }

        for (i in 0 until N)
            if (!chk[i]) {
                ++ans
                dfs(i)
            }

        bw.write("${ans}\n")
    }
}

fun dfs(i: Int) {
    if (chk[i]) return

    chk[i] = true

    for (j in 0 until N)
        if (gr[i][j] && !chk[j])
            dfs(j)
}

File: PART 2/Chapter 05/Problem 02/3.swift
import Foundation

var dic = [Int: Int]()
let _ = Int(readLine()!)
for i in readLine()!.split(separator: " ").map({ Int(String($0))! }) {
    if dic[i] == nil {
        dic[i] = 1
    } else {
        dic[i]! += 1
    }
}

let M = Int(readLine()!)!
var ans = [Int]()
for i in readLine()!.split(separator: " ").map({ Int(String($0))! }) {
    if dic[i] == nil {
        ans.append(0)
    } else {
        ans.append(dic[i]!)
    }
}

print(ans[0], terminator: "")
for i in 1..<M {
    print(" " + String(ans[i]), terminator: "")
}